[New Feature] Allowed mentions (#1396)
hammerandchisel/discord-api-docs: docs/resources/Channel.md

###### Allowed Mention Types

| Type              | Value      | Description                           |
| ----------------- | ---------- | ------------------------------------- |
| Role Mentions     | "roles"    | Controls role mentions                |
| User Mentions     | "users"    | Controls user mentions                |
| Everyone Mentions | "everyone" | Controls @everyone and @here mentions |

###### Allowed Mentions Structure

| Field | Type                           | Description                                                                                                   |
| ----- | ------------------------------ | ------------------------------------------------------------------------------------------------------------- |
| parse | array of allowed mention types | An array of [allowed mention types](#DOCS_RESOURCES_CHANNEL/allowed-mention-types) to parse from the content. |
| roles | list of snowflakes             | Array of role_ids to mention (Max size of 100)                                                                |

> | users | list of snowflakes             | Array of user_ids to mention (Max size of 100)                                                                |

Reviewer: Personal opinion here, probably subject to discussion, (and I recognize that this PR is already merged) but I think the `Allowed Mention Types` section should go *below* the `Allowed Mentions Structure` section, so that the user knows what the `Allowed Mention Types` section is referring to by the time they get to it. Currently, they read that section and aren't sure what to make of it, until they get to the `Allowed Mentions Structure` section.

It also just makes it more consistent with other parts of the documentation, like the [User Object section](https://discordapp.com/developers/docs/resources/user#user-object), where the tables describing enum-like things come *after* the tables that refer to them.

| Everyone Mentions | "everyone" | Controls @everyone and @here mentions |

###### Allowed Mentions Structure

| Field | Type                           | Description                                                                                                   |
| ----- | ------------------------------ | ------------------------------------------------------------------------------------------------------------- |
| parse | array of allowed mention types | An array of [allowed mention types](#DOCS_RESOURCES_CHANNEL/allowed-mention-types) to parse from the content. |
| roles | list of snowflakes             | Array of role_ids to mention                                                                                  |
| users | list of snowflakes             | Array of user_ids to mention                                                                                  |

###### Allowed Mentions Reference

Due to the complexity of possibilities, we have included a set of examples and behavior for the allowed mentions field.

If `allowed_mentions` is _not_ passed in (i.e. the key does not exist), the mentions will be parsed via the content.
This corresponds with existing behavior

> e.g. In the example below we would ping @here (and also @role124 and @user123)

Reviewer 2: ```suggestion
If `allowed_mentions` is _not_ passed in (i.e. the key does not exist), the mentions will be parsed via the content. This corresponds with existing behavior.

In the example below we would ping @here (and also @role124 and @user123)
```

spacing nits

```json
{
  "content": "@here Hi there from <@123>, cc <@&124>"
}
```

To suppress all mentions in a message use:

```json
{
  "content": "@everyone hi there, <@&123>",
  "allowed_mentions": {
    "parse": []
  }
}
```

This will suppress _all_ mentions in the message (no @everyone or user mention)

The `parse` field is mutually exclusive with the other fields
e.g. In the example below, we would ping users `123` and role `124`, but _not_ @everyone

> Note that passing a `Falsy` value ([], null) into the "users" field does not trigger a validation error.

Reviewer 2: ```suggestion
This will suppress _all_ mentions in the message (no @everyone or user mention).

The `parse` field is mutually exclusive with the other fields. In the example below, we would ping users `123` and role `124`, but _not_ @everyone. Note that passing a `Falsy` value ([], null) into the "users" field does not trigger a validation error.
```

```json
{
  "content": "@everyone <@123> <@&124>",
  "allowed_mentions": {
    "parse": ["users", "roles"],
    "users": []
  }
}
```

In the next example, we would ping @everyone, (and also users `123` and `124` if they suppressed
@everyone mentions), but we would not ping any roles.

```json
{
  "content": "@everyone <@123> <@124> <@125> <@&200>",
  "allowed_mentions": {
    "parse": ["everyone"],
    "users": ["123", "124"]
  }
}
```

Due to possible ambiguities, not all configurations are valid. An _invalid_ configuration is as follows

```json
{
  "content": "@everyone <@123> <@124> <@125> <@&200>",
  "allowed_mentions": {
    "parse": ["users"],
    "users": ["123", "124"]
  }
}
```

Because `parse: ["users"]` and `users: [123, 124]` are both present, we would throw a validation error.
This is because the conditions cannot be fulfilled simultaneously (they are mutually exclusive).

The ID list fields act as whitelists, and can contain mentions not in the content. These ID's that are not in the
content will simply be ignored.
e.g. The following example is valid, and would mention user 123, but _not_ user 125 since there is no mention of
user 125 in the content.

```json
{

>     "content": "<@123> Time for some memes."

Reviewer 3: nit: you missed a comma here
```suggestion
    "content": "<@123> Time for some memes.",
```

>     "content": "<@123> Time for some memes."

Reviewer 2: ```suggestion
    "content": "<@123> Time for some memes.",
```
